### PR TITLE
propagate "attention_mask" dtype for "use_past" in OnnxConfig.generate_dummy_inputs

### LINKED ISSUE #16538: ONNX causal-lm-with-past conversion: attention_mask dtype changed
## Environment info

- `transformers` version: 4.17.0
- Platform: Linux-5.4.144+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.13
- PyTorch version (GPU?): 1.10.0+cu111 (False)
- Tensorflow version (GPU?): 2.8.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: False
- Using distributed or parallel set-up in script?: Fasle

### Who can help

Not sure who should I tag for ONNX related issues @mfuntowicz ? tagging  @patil-suraj as contact for GPT models  

## Information

Model I am using : GPT2, GPTNeo

The problem arises when using the official conversion script (see below).

The tasks I am working on is pre-trained model conversion to ONNX

## To reproduce

1. Convert GPT model to ONNX for `causal-lm-with-past` using
```bash
python -m transformers.onnx --model=gpt2 --feature=causal-lm-with-past --atol=5e-4 ./onnx/
```
2. Load the model and check expected input types
```py
import onnx

model = onnx.load("onnx/model.onnx")
inp = model.graph.input
print(f"{inp[0].name}: element type {inp[0].type.tensor_type.elem_type}")
print(f"{inp[-1].name}: element type {inp[-1].type.tensor_type.elem_type}")
```
```
input_ids: element type 7
attention_mask: element type 1
```

The entire process can be reproduced using this colab: https://colab.research.google.com/gist/arampacha/2831d9f6812d2eb4d6d11dc13f76ca49/hf-onnx-attn-mask.ipynb
## Expected behavior

The `attention_mask` input should be of integer type (element type 7) not float (element type 1). Because of this `attention_mask` returned by tokenizer should be converted to `float` for inference.

Looks like the unexpected conversion happens because `torch.ones` returns `torch.float32` by default. See this for example https://github.com/huggingface/transformers/blob/9de70f213eb234522095cc9af7b2fac53afc2d87/src/transformers/models/gpt2/configuration_gpt2.py#L266

The problem can be fixed by propagating `attention_mask` dtype:
```py
mask_dtype = ordered_inputs["attention_mask"].dtype
ordered_inputs["attention_mask"] = torch.cat( 
    [ordered_inputs["attention_mask"], torch.ones(batch, past_key_values_length, dtype=mask_dtype)], dim=1
)
```
I can submit PR with a fix but more model classes can be impacted.

### PR DESCRIPTION
**Pull request by arampacha**
# What does this PR do?

Fixes #16538 

The `mask_dtype` is propagated to `torch.ones()` producing "attention_mask" for `past_key_values` in `generate_dummy_inputs` call. This ensures the input datatype expected by ONNX model matches default "attention_mask" `dtype`.
The fix is applied for configs where the pattern was used.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

The existing tests use `*OnnxConfig.generate_dummy_inputs` method to produce inputs passed to `session.run(...)` for testing. For this reason the issue was not reported at testing - the same inputs are used for export and testing. I'm not sure if specific tests for inputs datatype is required.

Here is a [notebook](https://colab.research.google.com/gist/arampacha/b123a1350223665201d3a3c6056cf490/hf_onnx_inference.ipynb) for verifying the fix works as expected.

## Who can review?

@lewtun

### CODE REVIEW

**Review by lewtun — APPROVED** [review]
Thanks for improving all these types for the attention masks @arampacha! 

It looks good to me, so gently pinging @patil-suraj and @sgugger for their perspective :)

For context to the reviewers: these changes ensure the data types of `input_ids` and `attention_mask` are the same (i.e. ints) when these models are exported to ONNX

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

**Review by patil-suraj — APPROVED** [review]
Thanks for the fix!
